src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+        /// within the schema constructor.
+        /// </remarks>
+        public static IGraphQLBuilder AddClrTypeMappings(this IGraphQLBuilder builder)
+            => builder.AddClrTypeMappings(Assembly.GetCallingAssembly());
+
+        /// <summary>
+        /// Scans the specified assembly for classes that inherit from <see cref="ObjectGraphType{TSourceType}"/>,
+        /// <see cref="InputObjectGraphType{TSourceType}"/>, or <see cref="EnumerationGraphType{TEnum}"/>, and
+        /// registers clr type mappings on the schema between that class and the source type or underlying enum type.
+        /// Skips classes where the source type is <see cref="object"/>, or where the class is marked with
+        /// the <see cref="DoNotMapClrTypeAttribute"/>.
+        /// </summary>
+        /// <remarks>
+        /// This is equivalent to calling <see cref="SchemaExtensions.RegisterTypeMappings(ISchema, Assembly)"/>
+        /// within the schema constructor.
+        /// </remarks>
+        public static IGraphQLBuilder AddClrTypeMappings(this IGraphQLBuilder builder, Assembly assembly)
+        {
+            if (assembly == null)
+                throw new ArgumentNullException(nameof(assembly));
+
+            // retreive all of the type mappings ahead-of-time, in case of a scoped or transient schema,
+            // as reflection is relatively slow
+            var typeMappings = assembly.GetClrTypeMappings();
+            builder.ConfigureSchema((schema, serviceProvider) =>
+            {
+                foreach (var typeMapping in typeMappings)
+                {
+                    schema.RegisterTypeMapping(typeMapping.ClrType, typeMapping.GraphType);
+                }
+            });
+
+            return builder;
+        }
+        #endregion
+
+        #region - AddDocumentListener -
+        /// <summary>
+        /// Registers <typeparamref name="TDocumentListener"/> with the dependency injection framework as both <typeparamref name="TDocumentListener"/> and
+        /// <see cref="IDocumentExecutionListener"/>. Configures document execution to add an instance of <typeparamref name="TDocumentListener"/> to the
+        /// list of document execution listeners within <see cref="ExecutionOptions.Listeners"/>. Singleton, scoped and transient lifetimes are supported.
+        /// </summary>
+        /// <remarks>
+        /// Do not separately add the document listener to your execution code or the document listener may be registered twice for the same execution.
+        /// </remarks>
+        public static IGraphQLBuilder AddDocumentListener<TDocumentListener>(this IGraphQLBuilder builder, ServiceLifetime serviceLifetime = ServiceLifetime.Singleton)
+            where TDocumentListener : class, IDocumentExecutionListener
+        {
+            builder.Register<TDocumentListener>(serviceLifetime);
+            builder.Register<IDocumentExecutionListener, TDocumentListener>(serviceLifetime);
```

Comment:
The same bug here:
```
 builder.Register<IDocumentExecutionListener, TDocumentListener>(serviceLifetime);
```
->
```
 builder.Register<IDocumentExecutionListener>(provider => provider.GetRequiredService<TDocumentListener>(), serviceLifetime);
```

Reply:
Same discussion about double disposal.

Reply:
Double disposal is not as terrible as a double creation. `Dispose` **should** support repeated calls, it is written everywhere.